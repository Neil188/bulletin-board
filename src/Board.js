import React, { Component } from 'react';
import PropTypes from 'prop-types';
import FaPlus from 'react-icons/lib/fa/plus';
import Note from './Note';
import defaultNoteData from './data.json';

export const nextId = ( prev ) => {
    const next = (prev || 0) + 1;
    return next;
};

export const addToArray = (value, id) => ({ notes=[] }) => ({
    notes: [
        ...notes,
        {
            id,
            note: value,
        },
    ],
});

export const removeFromArray = (id) => ({ notes=[] }) => ({
    notes: notes.filter(note =>
        note.id !== id
    ),
});

export const updateArray = (newText, i) => ({ notes=[] }) => ({
    notes: notes.map(
        note => note.id !== i ?
            note
            : {...note, note: newText}
    ),
});

export const getRandomNotes = ( count , { noteData }) => {
    let result = [];
    for (let x=count; x>0; x--) {
        const random = Math.floor(Math.random() * noteData.length);
        result = [...result, noteData[random] ];
    }
    return result;
};

export default class Board extends Component {

    state = {
        notes: [],
    };

    componentDidMount() {
        try {
            const retrieveNotes =
                JSON.parse(localStorage.getItem('bulletin-board'));

            if (retrieveNotes) {
                retrieveNotes.forEach( x =>
                    this.add(x.note)
                );

                return;
            }

            if (defaultNoteData) {
                getRandomNotes(5,defaultNoteData).forEach( x =>
                    this.add(x)
                );
            }
        } catch (e) {
            console.error('some kind of error ', e);
        }

    }

    componentDidUpdate() {
        const { notes } = this.state;
        const json = JSON.stringify(notes);
        localStorage.setItem('bulletin-board', json);
    }


    add = (text) => {
        this.uniqueId = nextId(this.uniqueId);
        const addNote = addToArray(text, this.uniqueId);
        this.setState(addNote);
    }

    addNew = () => this.add('New Note')

    eachNote = (note,i) => (
        <Note
            key={note.id}
            index={note.id}
            onChange={this.update}
            onRemove={this.remove}
            test={this.props.test}
        >
            {note.note}
        </Note>
    )

    remove = (id) => {
        const removeNote = removeFromArray(id);
        this.setState(removeNote);
    }

    update = (newText, i) => {
        const updateNote = updateArray(newText, i);
        this.setState( updateNote );
    }

    render() {
        return (
            <div className='board'>
                {this.state.notes.map(this.eachNote)}
                <button
                    onClick={this.addNew}
                    id='add'
                    aria-label='add'
                >
                    <FaPlus />
                </button>
            </div>
        );
    }
}

Board.defaultProps = {
    test: false,
};

Board.propTypes = {
    test: PropTypes.bool,
};
