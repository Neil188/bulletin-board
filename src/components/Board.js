import React, { Component } from 'react';
import FaPlus from 'react-icons/lib/fa/plus';
import FaTrash from 'react-icons/lib/fa/trash';
import Note from './Note';
import defaultNoteData from '../data/data.json';
import { getRandomNotes } from '../utils/random';

export const nextId = ( prev ) => {
    const next = (prev || 0) + 1;
    return next;
};

export const addToArray = (value, id) => ({ notes=[] }) => ({
    notes: [
        ...notes,
        {
            id,
            note: value,
        },
    ],
});

export const removeFromArray = (id) => ({ notes=[] }) => ({
    notes: notes.filter(note =>
        note.id !== id
    ),
});

export const updateArray = (newText, i) => ({ notes=[] }) => ({
    notes: notes.map(
        note => note.id !== i ?
            note
            : {...note, note: newText}
    ),
});

export default class Board extends Component {

    state = {
        notes: [],
    };

    componentDidMount() {
        try {
            const retrieveNotes =
                JSON.parse(localStorage.getItem('bulletin-board'));

            if (retrieveNotes) {
                retrieveNotes.forEach( x =>
                    this.add(x.note)
                );

                return;
            }

            if (defaultNoteData) {
                getRandomNotes(5,defaultNoteData).forEach( x =>
                    this.add(x)
                );
            }
        } catch (e) {
            console.error('some kind of error ', e);
        }

    }

    componentDidUpdate() {
        const { notes } = this.state;
        const json = JSON.stringify(notes);
        localStorage.setItem('bulletin-board', json);
    }


    add = (text) => {
        this.uniqueId = nextId(this.uniqueId);
        const addNote = addToArray(text, this.uniqueId);
        this.setState(addNote);
    }

    addNew = () => this.add('New Note')

    removeAll = () => this.setState( () => ({notes: []}))

    eachNote = (note,i) => (
        <Note
            key={note.id}
            index={note.id}
            onChange={this.update}
            onRemove={this.remove}
        >
            {note.note}
        </Note>
    )

    remove = (id) => {
        const removeNote = removeFromArray(id);
        this.setState(removeNote);
    }

    update = (newText, i) => {
        const updateNote = updateArray(newText, i);
        this.setState( updateNote );
    }

    render() {
        return (
            <div className='board'>
                {this.state.notes.map(this.eachNote)}
                <button
                    onClick={this.addNew}
                    id='add'
                    aria-label='add'
                >
                    <FaPlus />
                </button>
                <button
                    onClick={this.removeAll}
                    id='remove-all'
                    aria-label='remove all'
                >
                    <FaTrash />
                </button>
            </div>
        );
    }
}
