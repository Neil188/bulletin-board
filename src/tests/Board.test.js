import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Board, {
    nextId, addToArray, removeFromArray, updateArray,
} from '../components/Board';
import testData from '../data/data.json';

Enzyme.configure({
    adapter: new Adapter(),
});

jest.mock('../utils/random');

describe('function tests', () => {
    describe('Next Id tests', () => {
        test('prev = Undefined', () => {
            expect(nextId()).toBe(1);
        });

        test('prev = 1', () => {
            expect(nextId(1)).toBe(2);
        });
    });

    describe('Add to Array tests', () => {

        test('Add to empty array', () => {
            const result = addToArray('Test1', 1)({});
            expect(result.notes).toHaveLength(1);
            expect(result.notes[0].note).toBe('Test1');
        });

        test('Add to populated array', () => {
            const prev = {notes:[{id:1, note:'Test1'}]};
            const result = addToArray('Test2', 2)(prev);
            expect(result.notes).toHaveLength(2);
            expect(result.notes[1].note).toBe('Test2');
        });
    });

    describe('Remove Array tests', () => {

        test('Remove from empty array', () => {
            const prev = {};
            const result = removeFromArray(1)(prev);
            expect(result.notes).toHaveLength(0);
        });

        test('Remove from populated array (id found)', () => {
            const prev = {notes:[{id:1, note:'Test1'}]};
            const result = removeFromArray(1)(prev);
            expect(result.notes).toHaveLength(0);
        });

        test('Remove from populated array (id not found)', () => {
            const prev = {notes:[{id:1, note:'Test1'}]};
            const result = removeFromArray(2)(prev);
            expect(result.notes).toHaveLength(1);
        });
    });

    describe('Update Array tests', () => {

        test('Update empty array', () => {
            const prev = {};
            const result = updateArray('Test', 1)(prev);
            expect(result.notes).toHaveLength(0);
        });

        test('Update populated array (id found)', () => {
            const prev = {notes:[{id:1, note:'Test1'}]};
            const result = updateArray('Update', 1)(prev);
            expect(result.notes).toHaveLength(1);
            expect(result.notes[0].note).toBe('Update');
        });

        test('Update populated array (id not found)', () => {
            const prev = {notes:[{id:1, note:'Not changed'}]};
            const result = updateArray('Update', 2)(prev);
            expect(result.notes).toHaveLength(1);
            expect(result.notes[0].note).toBe('Not changed');
        });
    });

});

class LocalStorageMock {
    constructor() {
        this.store = {};
    }
    clear() {
        this.store = {};
    }
    getItem(key) {
        return this.store[key] || null;
    }
    setItem(key, value) {
        this.store[key] = value;
    }
}
global.localStorage = new LocalStorageMock;

describe('Board snapshots with no local storage', () => {
    test('Board snapshot test', () => {
        const wrapper = shallow(
            <Board />,
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('Validate note data', () => {
        const wrapper = shallow(
            <Board />,
        );
        expect(wrapper.state('notes')).toHaveLength(5);
        expect(wrapper.state('notes')[0]).toEqual({
            id: 1,
            note: testData.noteData[0],
        });
    });
});

describe('With mock local storage', () => {
    const testNotes = [
        {
            "id":1,
            note:"test",
        },
        {
            "id":2,
            note:"New Note",
        },
    ];
    let wrapper;

    beforeEach( () => {
        localStorage.setItem(
            'bulletin-board',
            JSON.stringify([testNotes[0]])
        );
        wrapper = shallow(
            <Board />,
        );
    });

    test('Board snapshot test', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('validate note data', () => {
        expect(wrapper.state('notes')).toHaveLength(1);
        expect(wrapper.state('notes')[0]).toEqual(testNotes[0]);
    });

    test('test buttons', () => {
        test('Add note', () => {
            wrapper.find('button').at(0).simulate('click');
            expect(wrapper.state('notes')).toHaveLength(2);
            expect(wrapper.state('notes')).toEqual(testNotes);
        });

        test('Remove all notes', () => {
            wrapper.find('button').at(1).simulate('click');
            expect(wrapper.state('notes')).toHaveLength(0);
        });
    });
});
