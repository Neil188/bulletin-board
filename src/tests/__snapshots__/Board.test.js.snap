// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board snapshots with no local storage Board snapshot test 1`] = `
<div
  className="board"
>
  <Note
    index={1}
    key="1"
    onChange={[Function]}
    onRemove={[Function]}
  >
    hello
  </Note>
  <Note
    index={2}
    key="2"
    onChange={[Function]}
    onRemove={[Function]}
  >
    hi
  </Note>
  <Note
    index={3}
    key="3"
    onChange={[Function]}
    onRemove={[Function]}
  >
    goedendag
  </Note>
  <Note
    index={4}
    key="4"
    onChange={[Function]}
    onRemove={[Function]}
  >
    guten tag
  </Note>
  <Note
    index={5}
    key="5"
    onChange={[Function]}
    onRemove={[Function]}
  >
    shalom
  </Note>
  <button
    aria-label="add"
    id="add"
    onClick={[Function]}
  >
    <FaPlus />
  </button>
  <button
    aria-label="remove all"
    id="remove-all"
    onClick={[Function]}
  >
    <FaTrash />
  </button>
</div>
`;

exports[`With mock local storage Add note 1`] = `
<div
  className="board"
>
  <Note
    index={1}
    key="1"
    onChange={[Function]}
    onRemove={[Function]}
  >
    test
  </Note>
  <Note
    index={2}
    key="2"
    onChange={[Function]}
    onRemove={[Function]}
  >
    New Note
  </Note>
  <button
    aria-label="add"
    id="add"
    onClick={[Function]}
  >
    <FaPlus />
  </button>
  <button
    aria-label="remove all"
    id="remove-all"
    onClick={[Function]}
  >
    <FaTrash />
  </button>
</div>
`;

exports[`With mock local storage Board snapshot test 1`] = `
<div
  className="board"
>
  <Note
    index={1}
    key="1"
    onChange={[Function]}
    onRemove={[Function]}
  >
    test
  </Note>
  <button
    aria-label="add"
    id="add"
    onClick={[Function]}
  >
    <FaPlus />
  </button>
  <button
    aria-label="remove all"
    id="remove-all"
    onClick={[Function]}
  >
    <FaTrash />
  </button>
</div>
`;

exports[`With mock local storage Remove all notes 1`] = `
<div
  className="board"
>
  <button
    aria-label="add"
    id="add"
    onClick={[Function]}
  >
    <FaPlus />
  </button>
  <button
    aria-label="remove all"
    id="remove-all"
    onClick={[Function]}
  >
    <FaTrash />
  </button>
</div>
`;

exports[`With mock local storage Remove note 1`] = `
<div
  className="board"
>
  <Note
    index={2}
    key="2"
    onChange={[Function]}
    onRemove={[Function]}
  >
    New Note
  </Note>
  <button
    aria-label="add"
    id="add"
    onClick={[Function]}
  >
    <FaPlus />
  </button>
  <button
    aria-label="remove all"
    id="remove-all"
    onClick={[Function]}
  >
    <FaTrash />
  </button>
</div>
`;

exports[`With mock local storage Update note 1`] = `
<div
  className="board"
>
  <Note
    index={1}
    key="1"
    onChange={[Function]}
    onRemove={[Function]}
  >
    New text
  </Note>
  <button
    aria-label="add"
    id="add"
    onClick={[Function]}
  >
    <FaPlus />
  </button>
  <button
    aria-label="remove all"
    id="remove-all"
    onClick={[Function]}
  >
    <FaTrash />
  </button>
</div>
`;
